### ISSUE TITLE
Access URL Map

### ISSUE THREAD

**Issue by foggy1** [bitrise-steplib/steps-deploy-to-bitrise-io#59]
I'm working with a react native app, and I'm able to get everything to build, but I'd like to send the full install list map (with the android AND ios links) in one single go to Slack.

When I try to do this right now, I do it as such within the table `All Install Links|${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP}`

This works, but it dumps out the map as one single blob of text (e.g. "app-release.apk=>https://www.bitrise.io/artifact/blahhhhhhapp-release20180315133216.apk=>https://www.bitrise.io/artifact/blahhhhhhhhhapp.ipa=>https://www.bitrise.io/artifact/blahhhhhh"). And it looks just like this one, where it partially links and they're not separated and it's a pain to partially select and copy paste, etc.

I have very little experience with Go but I'm assuming I could write a Go script to map through this, break it up, and assign it to piecemeal ENV vars for more efficient use. But am I overcomplicating things? 

**In short**: I'm wondering if there's some way for me to access the keys in this map straight from the point in the Slack integration where I invoke the ENV var, OR if I have to write a Go script to break it up by value right now.

EDIT: Currently I have a Ruby script that's regex-ing this field and creating new ENV variables with the value at each key. This is... fine, but a little silly that I have to do this.

**Comment by viktorbenei** [comment]
> EDIT: Currently I have a Ruby script that's regex-ing this field and creating new ENV variables with the value at each key. This is... fine, but a little silly that I have to do this.

That sounds a reasonable solution.

Ideally what would you like to see / how should it work?

**Comment by foggy1** [comment]
If the urls are available in order to generate a map, it seems like it would be pretty cheap to just assign them all to individual ENV variables to start with. One of the biggest upsides of Bitrise is how easy it is to drop in ENV variables from throughout the flow.

Dropping in a whole map, however, is way more costly, as its implementation and the rules for dealing with it vary throughout contexts, whereas a single ENV wrapping a single url can be dropped in anywhere with the same expected result.

I don't know up from down when it comes to Go, but it looks like the current script indiscriminately maps through all of the install pages in order to create the map. Earlier, however, it differentiates between the *.ipa and the *.apk. In those branches specifically the install pages could be mapped to a specific ENV variables.

Of course, now looking at this, I appreciate that the appeal of the map is 1. it doesn't care what kind of or how many packages you built and 2. the approach I'm suggesting, with enough packages being built, would lead to slightly unpredictable ENV variables. So, I don't know if my suggestion is even practical!

**Comment by viktorbenei** [comment]
> If the urls are available in order to generate a map, it seems like it would be pretty cheap to just assign them all to individual ENV variables to start with

Definitely, but what would be the ENV key? What would it based on?

In general our recommendation if you want the URLs separately, and the map doesn't work, to add as many Deploy to Bitrise.io steps as many IPAs and APKs you want to deploy, and simply set the path one by one. That way every Deploy to Bitrise.io step will expose the single Public Install Page URL in the `BITRISE_PUBLIC_INSTALL_PAGE_URL` output env var (https://github.com/bitrise-io/steps-deploy-to-bitrise-io/blob/b976ce63b35e6174fa1b27b9d4971faeade5d68b/step.yml#L148), which you can alias to any other env var (https://discuss.bitrise.io/t/monthly-release-of-bitrise-cli-tools-and-summary-of-updates-may-2017-cli-v1-6-1/1690 -> **Step Output Aliases** section).

**Comment by darkplataform** [comment]
Hi, I have the very same problem, but I only work in an android project which have two flavors, and got two apk for each flavor(1 for release and 1 for debug).  The map is generated correctly and assigned to BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP variable:

BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP (value: XXXXXX-OtherRelease-1.6.4-1000175.apk=>https://www.bitrise.io/artifact/xxxxxx/p/yyyyyyy |XXXXXX-YYYYYYYYDebug-1.6.4-1000175.apk=>https://www.bitrise.io/artifact/xxxxxx/p/yyyyyyy|XXXXXX-YYYYYYYYRelease-1.6.4-1000175.apk=>https://www.bitrise.io/artifact/xxxxxx/p/yyyyyyy|XXXXXX-OtherDebug-1.6.4-1000175.apk=>https://www.bitrise.io/artifact/xxxxxx/p/yyyyyyy)

However in slack I only get these three buttons(The install page by the way only aims one artifact):
![image](https://user-images.githubusercontent.com/5014699/38321766-18669b34-37fe-11e8-8b1c-9dedb18a91c3.png)

I dont even get the blob of  a text that @foggy1  is mentioning. And its correctly mapped in the Slack Step:

`View App|${BITRISE_APP_URL}`
`View Build|${BITRISE_BUILD_URL}`
`Install Page|${BITRISE_PUBLIC_INSTALL_PAGE_URL}`
`Install Map|${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP}`

I don't understand why it wont work if the MAP variable is correctly set.

> to add as many Deploy to Bitrise.io steps as many IPAs and APKs you want to deploy, and simply set the path one by one

I understand what your saying but I think it breaks the nice automated feature of "Deploy to Bitrise Step" that take care of all your artifacts/flavors/etc without much or any customization, also I personally think it would be nice to add a flavor to my project and forget about updating bitrise.

**Comment by viktorbenei** [comment]
@darkplataform can you please create a separate issue for this problem, with the link of the build or full log? It seems yours is a separate issue and we'll need more info (e.g. the build URL) to be able to tell what the issue is there (most likely a config issue, but can't tell anything useful without a URL :))

**Comment by darkplataform** [comment]
@viktorbenei  I submit a new issue, but I manage to get the blob text putting the env variable on the table of attachments, please see the issue I opened:
https://github.com/bitrise-io/steps-slack-message/issues/38

**Comment by gitizenme** [comment]
The blob ends up rendering as on massive link, see: 

![image](https://user-images.githubusercontent.com/643504/38513336-c3958212-3be2-11e8-92f4-557ba7449da5.png)

**Comment by darkplataform** [comment]
That is exactly what I did, forgot to share it! thanks for remind me 👍 : 
```
#bash step
newline=" "
BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP_TEMP=${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP//|/$newline}
BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP_TEMP=${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP_TEMP//=>/ => }
envman add --key BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP --value "$BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP_TEMP"
```

**Comment by ValeryP** [comment]
@darkplataform haha.. synergy 💪 
Would be also nice if you can share this part of the workflow `bitrise.yml` config to understand where to place it correctly.

**Comment by darkplataform** [comment]
@ValeryP good call!:

workflows:
  QA:
    steps:
    - script@1.1.5:
        inputs:
        - content: "#some validations"
        title: Validate Build
    - build-number-adjust@1.0.0:
        run_if: enveq "RUN_FULL_BUILD"  "0"
        inputs:
        - decrease: '1'
    - slack@2.7.2:
        run_if: enveq "RUN_FULL_BUILD"  "1"
        inputs:
        - text: A build just started in the Development Branch
        - buttons: "\n"
        - pretext: ''
        - webhook_url: https://hooks.slack.com/services/
        title: Commit received to Slack
    - activate-ssh-key@3.1.1:
        run_if: enveq "RUN_FULL_BUILD"  "1"
    - git-clone@4.0.10:
        run_if: enveq "RUN_FULL_BUILD"  "1"
    - install-missing-android-tools@2.1.1:
        run_if: enveq "RUN_FULL_BUILD"  "1"
    - gradle-runner@1.8.3:
        run_if: enveq "RUN_FULL_BUILD"  "1"
        inputs:
        - gradle_file: "$GRADLE_BUILD_FILE_PATH"
        - gradle_options: "--stacktrace -PversionCode=$BUILD_VERSION_CODE -PversionName=$BUILD_VERSION_NAME"
        - gradle_task: assemble lint $RUN_ANALYSIS
        - gradlew_path: "$GRADLEW_PATH"
    - deploy-to-bitrise-io@1.3.10:
        is_always_run: false
        run_if: enveq "RUN_FULL_BUILD"  "1"
        inputs:
        - zip_name: xxxxxxxx_Android_Artifacts_${BITRISE_BUILD_NUMBER}.zip
    - script@1.1.5:
        run_if: enveq "RUN_FULL_BUILD"  "1"
        title: Create Tag
        inputs:
        **- content: "newline=\" \"\nBITRISE_PUBLIC_INSTALL_PAGE_URL_MAP_TEMP=${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP//|/$newline}\nBITRISE_PUBLIC_INSTALL_PAGE_URL_MAP_TEMP=${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP_TEMP//=>/
            => }\nenvman add --key BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP --value \"$BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP_TEMP\""**
    - slack@2.7.2:
        run_if: (enveq "RUN_FULL_BUILD"  "0")|  and (enveq "BITRISE_BUILD_STATUS"
          "0") |  not
        inputs:
        - text: ''
        - text_on_error: Build failed please review the Bitrise QA workflow
        - buttons: |-
            View App|${BITRISE_APP_URL}
            View Build|${BITRISE_BUILD_URL}
            Install Map|${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP}
        - fields: |-
            App|${BITRISE_APP_TITLE}
            Branch|${BITRISE_GIT_BRANCH}
            Workflow|${BITRISE_TRIGGERED_WORKFLOW_ID}
            Install Map|${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP}
        - webhook_url: https://hooks.slack.com/services///
        title: Build finished Slack

**Comment by lszucs** [comment]
Hi @foggy1 

We released a new feature, which gives you the opportunity to format the map value using a Go template. The `BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP` variable will be populated with the output of the template.

You may provide the template in the `public_install_page_url_map_format` input.

For example:

`
        - public_install_page_url_map_format: |
            {{range $index, $element := .}}
                {{$element.File}} => {{$element.URL}}
            {{end}}
`

Let me know if this helps and if we can close the issue.

**Comment by WarrenFaith** [comment]
@lszucs this works for me, but please have a look at #104

**Comment by foggy1** [comment]
@lszucs @godrei it's been awhile since I touched bitrise but this seems like a reasonable solution since the crux of the issue was having to dig into code and grep stuff in an ad hoc manner. A template solves that for sure.

**Comment by morizotter** [comment]
@darkplataform 's solution is really good! But in my case the new line didn't work. So I write my shell script inspired by @darkplataform 's solution.

My solution is: 

Add Script with content below after `Deploy to Bitrise.io` .

```sh
LIST=(${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP//|/ })
INSTALL_PAGE_0=${LIST[0]//=>/|}
INSTALL_PAGE_1=${LIST[1]//=>/|}

envman add --key INSTALL_PAGE_BUTTON_0 --value "$INSTALL_PAGE_0"
envman add --key INSTALL_PAGE_BUTTON_1 --value "$INSTALL_PAGE_1"
```

And in the slack buttons content:
```sh
View App|${BITRISE_APP_URL}
View Build|${BITRISE_BUILD_URL}
${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP}
```

Finally run,  each public install page buttons will show on slack message!

**Comment by jonchui** [comment]
@morizotter and @darkplataform's solution did not work for me.

I had an iOS and ANDROID app built, and my `${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP}`

looks like:

```
app-debug.apk=>https://app.bitrise.io/artifact/62249510/p/[REDACTED]%7CRunner.ipa\=\>https://app.bitrise.io/artifact/62249530/p/[REDACTED]
```

So instead, I used output aliases in the `deploy to bitrise` step, and created 2 of them. one for iOS and one for Android, like so:

```
- deploy-to-bitrise-io@1:
        inputs:
        - notify_user_groups: owner
        outputs:
        - BITRISE_PUBLIC_INSTALL_PAGE_URL: APK_PUBLIC_INSTALL_PAGE
- deploy-to-bitrise-io@1:
        inputs:
        - deploy_path: $BITRISE_IPA_PATH
        - notify_user_groups: owner
        outputs:
        - BITRISE_PUBLIC_INSTALL_PAGE_URL: IPA_PUBLIC_INSTALL_PAGE
```

Then in my slack `buttons` :
```
- buttons: |
            View App|${BITRISE_APP_URL}
            View Build|${BITRISE_BUILD_URL}
            iOS App|${IPA_PUBLIC_INSTALL_PAGE}
            Android App|${APK_PUBLIC_INSTALL_PAGE}
```
